### PR TITLE
Mf4 io

### LINKED ISSUE #506: add MF4 io writer
I would like to implement support for exporting to asam MF4 files (https://www.asam.net/standards/detail/mdf/). That would require my library as dependency ([asammdf](https://github.com/danielhrisca/asammdf)). Is this acceptable?

### PR DESCRIPTION
**Pull request by driftregion**
I rebased https://github.com/hardbyte/python-can/pull/554/

Closes #506

### CODE REVIEW

**Comment by felixdivo** [comment]
[mf4-io](https://github.com/hardbyte/python-can/tree/mf4-io) can probably be deleted, should this be merged.

**Comment by zariiii9003** [comment]
@driftregion can i force push to this PR?
I fixed a few things [here](https://github.com/zariiii9003/python-can/tree/mf4-io). I can open a new PR, too.

**Comment by driftregion** [comment]
@zariiii9003 sure, go ahead.

**Comment by zariiii9003** [comment]
@danielhrisca would it be possible to stream to a file directly?  
Also: i accessed a private attribute to get the current file size, is there a better way to do that?

**Comment by danielhrisca** [comment]
> @danielhrisca would it be possible to stream to a file directly?

The only way to `stream` to the file is to use the `extend` method, but this is performance bottleneck for high data rates.

> Also: i accessed a private attribute to get the current file size, is there a better way to do that?
I don't think so. Why do you need that information?

**Comment by zariiii9003** [comment]
@danielhrisca 
The SizedRotatingLogger starts a new file when a give file size is reached. That's why i added the `file_size` function to the logger classes.

The BlfWriter compresses and writes data in chunks so maybe that could be possible here too?

File: setup.py [hunk 1]
```diff
@@ -36,6 +36,10 @@
     "viewer": [
         'windows-curses;platform_system=="Windows" and platform_python_implementation=="CPython"'
     ],
+    "mf4": [
+        'asammdf>=6.0.0;platform_python_implementation=="CPython" and python_version<"3.11"',
+        'numpy;platform_python_implementation=="CPython" and python_version<"3.11"',
```
Comment (hardbyte) [inline setup.py:41]: `asammdf` will declare the version of numpy it needs. Not sure about the platform and python version requirement either?

File: doc/listeners.rst [hunk 2]
```diff
@@ -191,3 +191,32 @@ The following class can be used to read messages from BLF file:
 
 .. autoclass:: can.BLFReader
     :members:
+    
+  
+MF4 (Measurement Data Format v4)
+--------------------------------
+
+Implements support for MF4 (Measurement Data Format v4) which is a proprietary
+format from ASAM, widely used in many automotive software (Vector CANape, ETAS INCA, dSPACE ControlDesk, etc.).
```
Comment (hardbyte) [inline doc/listeners.rst:200]:
```suggestion
format from ASAM (Association for Standardization of Automation and Measuring Systems), widely used in 
many automotive software (Vector CANape, ETAS INCA, dSPACE ControlDesk, etc.).
```

File: can/io/mf4.py [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+        self._std_buffer = np.zeros(1, dtype=STD_DTYPE)
+        self._err_buffer = np.zeros(1, dtype=ERR_DTYPE)
+        self._rtr_buffer = np.zeros(1, dtype=RTR_DTYPE)
+
+
+class MF4Reader(MessageReader):
+    """
+    Iterator of CAN messages from a MF4 logging file.
+
+    The MF4Reader only supports MF4 files that were recorded with python-can.
```
Comment (hardbyte) [inline can/io/mf4.py:264]: Why is that?

File: setup.py (hunk 1 above)
Comment (zariiii9003) [reply setup.py]: I think i added this so CI does not fail on PyPy and CPython 3.11. But there's probably a better solution for that. I admit this is ugly.

File: can/io/mf4.py (hunk 3 above)
Comment (zariiii9003) [reply can/io/mf4.py:264]: It's a limitation of asammdf. See [here](https://github.com/danielhrisca/asammdf/issues/744#issuecomment-1207942077)

**Comment by zariiii9003** [comment]
@danielhrisca merged after only 3.5 years 💪

**Comment by danielhrisca** [comment]
It's like a good wine

**Pull request merged**